File: src/leetcode/easy/ConvertSortedArrayToBinarySearchTree.java
package leetcode.easy;

import common.TreeNode;

/**
 * Question: https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/description/
 */
public class ConvertSortedArrayToBinarySearchTree {
    public TreeNode sortedArrayToBST(int[] nums) {
        return sortedArrayToBST(nums, 0, nums.length - 1);
    }

    private TreeNode sortedArrayToBST(int[] nums, int start, int end) {
        if (start > end) return null;
        int middle = (start + end) / 2;
        TreeNode node = new TreeNode(nums[middle]);
        node.left = sortedArrayToBST(nums, start, middle - 1);
        node.right = sortedArrayToBST(nums, middle + 1, end);
        return node;
    }
}

File: javascript/easy/convertSortedArrayToBinarySearchTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
var sortedArrayToBST = function(nums) {
    return constructBST(nums, 0, nums.length-1);
};

var constructBST = function(nums, start, end) {
    if(start > end) return null;
    let middle = parseInt((start + end)/2);
    let root = new TreeNode(nums[middle]);
    root.left = constructBST(nums, start, middle-1);
    root.right = constructBST(nums, middle+1, end);
    return root;
};
